### ISSUE TITLE
Different behavior between inline `enum` and out-of-line `enum`.

### ISSUE THREAD

**Issue by gregmarr** [go-swagger/go-swagger#1328]
## Problem statement

There is different behavior between inline `enum` and out-of-line `enum`.  The first creates a `*string` in the struct, and the second creates a go enum in the struct.  It would be nice if the first would also have the enum type in the struct.

## Swagger specification

    enum_type_one: { type: string, enum: [ '1' ] }

    included:
      type: array
      items:
        type: object
        properties:
          ida: { type: string, enum: [ '1' ] }
          idb: { $ref: '#/definitions/enum_type_one' }

The generated models:

    const (
	// CreateItemIncludedItemsIDANr1 captures enum value "1"
	CreateItemIncludedItemsIDANr1 string = "1"
    )

    // EnumTypeOne enum type one
    // swagger:model enum_type_one
    type EnumTypeOne string

    const (
	    // EnumTypeOneNr1 captures enum value "1"
	    EnumTypeOneNr1 EnumTypeOne = "1"
    )

    // for schema
    var enumTypeOneEnum []interface{}

	// ida
	// Required: true
	IDA *string `json:"ida"`

	// idb
	// Required: true
	IDB EnumTypeOne `json:"idb"`

## Environment
swagger version: master from Dec 11, 2017

**Comment by gregmarr** [comment]
I did Dump Data and looked at the containing structure.  Here are the significant differences:

```                        {
                            "Enum": [ "1" ],
                            "GoType": "string",
                        },
                        {
                            "AliasedType": "string",
                            "GoType": "EnumTypeOne",
                            "IsAliased": true
                        }
```
and `EnumTypeOne` gets its own file.
So it looks like we'd need to generate an aliased type here, or else do the transformation inside the template when creating the containing type.

**Comment by gregmarr** [comment]
Looks like a duplicate of #868

**Comment by fredbi** [comment]
Thank you greg for pointing that one out. I include it in my development test cases for rework of the enum const feature.

**Comment by fredbi** [comment]
@gregmarr : I think that I fixed this as an unexpected byproduct of my dev... I tried your test case on my WIP.
Isn't that what you would expect?
```golang
// IncludedItems included items
// swagger:model includedItems
type IncludedItems struct {

	// ida
	Ida string `json:"ida,omitempty"`

	// idb
	Idb EnumTypeOne `json:"idb,omitempty"`
}

// Validate validates this included items
func (m *IncludedItems) Validate(formats strfmt.Registry) error {
	var res []error

	if err := m.validateIda(formats); err != nil {
		// prop
		res = append(res, err)
	}

	if err := m.validateIdb(formats); err != nil {
		// prop
		res = append(res, err)
	}

	if len(res) > 0 {
		return errors.CompositeValidationError(res...)
	}
	return nil
}

const (
	// IncludedItemsIdaNr1 captures enum value 1
	IncludedItemsIdaNr1 = "1"
)

// for property

const idaEnum = []string{IncludedItemsIdaNr1}

func (m *IncludedItems) validateIdaEnum(path, location string, value string) error {

	if err := validate.Enum(path, location, value, idaEnum); err != nil {
		return err
	}
	return nil

}

func (m *IncludedItems) validateIda(formats strfmt.Registry) error {

	if swag.IsZero(m.Ida) { // not required
		return nil
	}

	// value enum
	if err := m.validateIdaEnum("ida", "body", m.Ida); err != nil {
		return err
	}

	return nil
}
```

**Comment by gregmarr** [comment]
Nope, that's the same as what we get now.  The `Ida` is just a `string` with runtime validation, not a separate type as `Idb` is.

**Comment by fredbi** [comment]
oh. I thought it was a missing validation problem. I got it now. There is all of a mumbo-jumbo in the model generation specifically to DO THIS. There is a reason why this "full aliasing" never completed: it is just highly risky to change it all. I am trying to carry on incremental changes on const generation, but I am afraid I'll step into all these problems as soon as I start testing nested structures... It is quite hard to solve/improve one thing at a time in schemavalidator.gotmpl... Any idea is welcome

**Comment by gregmarr** [comment]
I had some thoughts above, on Jan 8.  These are fairly uninformed thoughts.  I haven't looked at them to see how difficult they would be.

**Comment by casualjim** [comment]
we really need flatten but the changes to $ref resolution broke flatten too
Once you flatten these types of enums can be lifted to top level enums and everything will just work.
It's hard to solve this in a way that doesn't end up with very long names

**Comment by fredbi** [comment]
yeah. I figured out that we could work around this by allowing for custom names... But it's deeper.

customization works on my bench, and it allowed me to do what I wanted: reusable consts with any time (date, etc..) + fixing simple bugs because of funny values with no possible practical naming inference.

But I am afraid of entering the testing stage with all the nested stuff and additionalProperties and all : several things don't currently work and adding a _tested_ feature is much like you fix them all...

**Comment by fredbi** [comment]
@gregmarr I thought about this issue and it just occured to me that we cannot easily adopt a general approach to this.

In your example, you explicitly defined a `$ref` ,which is interpreted as "I want a specific type for that" (e.g. somehing that I may reuse). When defining the `enum` inline, there is no such apparent intent.

So it is not obvious that "lifting things" is always the right choice. Indeed, if this enum type is reused elsewhere, inlining it would cause name conflicts with the generated consts.

This kind of fix definitely lies in the `flatten()` func (github.com/go-openapi/analysis).
You probably know that there is a specific section there for this kind of additional rule (none is implemented yet).

I don't think we can adopt a general rule on simple types: some dev would like inlined enum to remain inline, and some, to have it as dedicated type.

We could instruct the flattener with some kind of `x-...` extension. At the moment,
the easiest way to express commonality is to use `$ref`.

On the other hand, you might have expected that the analyzer automatically detects that the schemas are actually identical and factorize the spec for you. But that does sound a bit far fetched, doesn't it?

Cheers,

F.

**Comment by gregmarr** [comment]
The big difference here is runtime validation vs compile time validation.  I'm not as concerned about commonality as the compile time validation.  

>So it is not obvious that "lifting things" is always the right choice. Indeed, if this enum type is reused elsewhere, inlining it would cause name conflicts with the generated consts.

I don't agree.  We already generate the consts for runtime validation, we just don't require that you use them when creating objects in code.

>I don't think we can adopt a general rule on simple types: some dev would like inlined enum to remain inline, and some, to have it as dedicated type.

I can't imagine devs wanting less type safety.  (Okay, I can imagine it, I just can't understand it.)

>On the other hand, you might have expected that the analyzer automatically detects that the schemas are actually identical and factorize the spec for you.

Nope, not looking for that.

**Comment by fredbi** [comment]
Hello @gregmarr 

> The big difference here is runtime validation vs compile time validation. 

Could you please be more explicit about this statement? You lost me. I don't understand what you mean by "compile time validation" in this context.

> I can't imagine devs wanting less type safety

Again I don't understand what type safety has to do with inlining.
In both cases, we have a validation on the enum, have we not?
Why having a dedicated type should be considered _safer_?


What I am getting at is that I don't really see why specifically enum should _automatically_ trigger the construction of a dedicated type (btw, a trick that would not be available with simple schemas),
when it is so straightforward to put this intent in the spec with a $ref ?

If this is what you want, I think it is rather straightforward (you already did tamper with that code, didn't you?) to set an option in flatten() to move all enums together with other complex stuff and create a definition for it. I guess that this would happen [here](https://github.com/go-openapi/analysis/blob/863ac7f90e00e88e507095639a8457bbbf3c2ec9/flatten.go#L91).

Maybe part of my misunderstanding comes from different use cases: I am _writing_ specs - so I change them until the model suits me. I realize you might need to rework models on existing specs without altering them.

**Comment by gregmarr** [comment]
```// ida
// Required: true
IDA *string `json:"ida"`

// idb
// Required: true
IDB EnumTypeOne `json:"idb"```

**Comment by gregmarr** [comment]
Here I'm copying from #868 

Enums occur in following locations, and should generate type aliases in all of them:

non-body parameter:
  inline (1);
  ref (2);
header object (used to send headers along with responses);
  inline (3);
items object (used to describe array items):
  inline (4);
  ref (5) (already working pre-PR as a solution to this issue!);

**Comment by fredbi** [comment]
Precisely, I am challenging the relevance of implementing (1), (3) and (4) in codegen.

In my view, this is spec massaging before generation. Flatten is definitely the place for such rework, that should be optional, so as not to break things.

**Comment by gregmarr** [comment]
I don't think I should need to put anything in the spec to get consistent behavior between the two definitions.  The spec author has specified that only the given values are valid for that property, so we should do everything that we can to make it easy for the user of the generated code to make that happen.  Having to add `$ref` and an external declaration for something that isn't reusable just because otherwise the tool doesn't give me the full compile time validation just feels wrong to me.  If we didn't have the ability to do it at all, that would be one thing, but we do, and we're just choosing not to do it in this case.

Something to remember here is that the use case where I hit this is in client generation.  I'm using a spec provided to me by a third party, and that third party doesn't use a separate type when the enum isn't shared.  This results in an inconsistent interface for the user of the client.  It's not always possible to make these changes to the spec to get the consistent interface.

Since it is a breaking change, a change like this would probably need to be part of the next major release, and I'm fine with that.

**Comment by fredbi** [comment]
@gregmarr Here is a new proposal, following what we discussed.Tell me if this is more like what you would expect.

This feature would be optional: either you enable it globally from a CLI option, or locally, with an `x-go-enum-type` vendor extension. It works for all inlined primitive properties with an enum.

Here is the example I have generate on dev from your spec (modified below, but again the vendor ext is not mandatory):
```yaml 
definitions:
  enum_type_one: { type: string, enum: [ '1' ] }

  included:
    type: array
    items:
      type: object
      properties:
        ida: { type: string, enum: [ '1' ], x-go-enum-type: true }
        idb: { $ref: '#/definitions/enum_type_one' }
```
As usual, you get the aliased type definition in its own file (note that we don't have init() and unmarshalling anymore here):
```go
// EnumTypeOne enum type one
// swagger:model enum_type_one
type EnumTypeOne string

// enum for schema enum_type_one

const (
	// EnumTypeOneNr1 captures enum value "1"
	EnumTypeOneNr1 = EnumTypeOne("1")
)

// enumTypeOneEnum represents the list of allowed enum values for this enum type one
var enumTypeOneEnum = []EnumTypeOne{EnumTypeOneNr1}

// validateEnumTypeOneEnum validates against enum for enum type one
func (m EnumTypeOne) validateEnumTypeOneEnum(path, location string, value EnumTypeOne) error {
	if err := validate.Enum(path, location, value, enumTypeOneEnum); err != nil {
		return err
	}
	return nil
}

// Validate validates this enum type one
func (m EnumTypeOne) Validate(formats strfmt.Registry) error {
	var res []error

	// primitive enum
	if err := m.validateEnumTypeOneEnum("", "body", m); err != nil {
		return err
	}

	if len(res) > 0 {
		return errors.CompositeValidationError(res...)
	}
	return nil
}
```

Then you get your inline type as extra-schema for your items. It is in the same file as the original type it stems from.

This works in both flatten and `--skip-flatten` modes and does not depend on analysis' `flatten()`:
```go
// IncludedItems included items
// swagger:model includedItems
type IncludedItems struct {

	// ida
	Ida IncludedItemsIdaEnum `json:"ida,omitempty"`

	// idb
	Idb EnumTypeOne `json:"idb,omitempty"`
}

// Validate validates this included items
func (m *IncludedItems) Validate(formats strfmt.Registry) error {
	var res []error

	if err := m.validateIda(formats); err != nil {
		res = append(res, err)
	}

	if err := m.validateIdb(formats); err != nil {
		res = append(res, err)
	}

	if len(res) > 0 {
		return errors.CompositeValidationError(res...)
	}
	return nil
}

func (m *IncludedItems) validateIda(formats strfmt.Registry) error {

	if swag.IsZero(m.Ida) { // not required
		return nil
	}

	if err := m.Ida.Validate(formats); err != nil {
		if ve, ok := err.(*errors.Validation); ok {
			return ve.ValidateName("ida")
		}
		return err
	}

	return nil
}

func (m *IncludedItems) validateIdb(formats strfmt.Registry) error {

	if swag.IsZero(m.Idb) { // not required
		return nil
	}

	if err := m.Idb.Validate(formats); err != nil {
		if ve, ok := err.(*errors.Validation); ok {
			return ve.ValidateName("idb")
		}
		return err
	}

	return nil
}

// MarshalBinary interface implementation
func (m *IncludedItems) MarshalBinary() ([]byte, error) {
	if m == nil {
		return nil, nil
	}
	return swag.WriteJSON(m)
}

// UnmarshalBinary interface implementation
func (m *IncludedItems) UnmarshalBinary(b []byte) error {
	var res IncludedItems
	if err := swag.ReadJSON(b, &res); err != nil {
		return err
	}
	*m = res
	return nil
}

// IncludedItemsIdaEnum included items ida enum
// swagger:model IncludedItemsIdaEnum
type IncludedItemsIdaEnum string

// enum for schema IncludedItemsIdaEnum

const (
	// IncludedItemsIdaEnumNr1 captures enum value "1"
	IncludedItemsIdaEnumNr1 = IncludedItemsIdaEnum("1")
)

// includedItemsIdaEnumEnum represents the list of allowed enum values for this included items ida enum
var includedItemsIdaEnumEnum = []IncludedItemsIdaEnum{IncludedItemsIdaEnumNr1}

// validateIncludedItemsIdaEnumEnum validates against enum for included items ida enum
func (m IncludedItemsIdaEnum) validateIncludedItemsIdaEnumEnum(path, location string, value IncludedItemsIdaEnum) error {
	if err := validate.Enum(path, location, value, includedItemsIdaEnumEnum); err != nil {
		return err
	}
	return nil
}

// Validate validates this included items ida enum
func (m IncludedItemsIdaEnum) Validate(formats strfmt.Registry) error {
	var res []error

	// primitive enum
	if err := m.validateIncludedItemsIdaEnumEnum("", "body", m); err != nil {
		return err
	}

	if len(res) > 0 {
		return errors.CompositeValidationError(res...)
	}
	return nil
}
```

**Comment by casualjim** [comment]
will this continue to chain names with nested anonymous schemas?
I don't know how much this is helping vs getting people to just define their own constants. These enums are typically such small things. While the code being generated here it probably doesn't actually capture intent.
The intent being typesafe reuse, and unless we analyze all the enums and group the ones that are the same we don't capture that intent.

**Comment by casualjim** [comment]
will this continue to chain names with nested anonymous schemas?
I don't know how much this is helping vs getting people to just define their own constants. These enums are typically such small things. 

While the code being generated here is valid, compiles and works, it probably doesn't actually capture intent. The intent being typesafe reuse, and unless we analyze all the enums and group the ones that are the same we don't capture that intent.

**Comment by fredbi** [comment]
This new schema node is terminal: it just applies to primitives. So there is no further chaining down.

Sure this is just a nice to have - we agreed on that with Greg. It did not take much time to develop, since the enum stuff is refactored on my dev branch.

I think it captures the implied intent with enum to define a safe enumerated type, but does not go so far as capturing the intent that this one is actually the same as the other one.

Anyway, I made it optional, so there is no break to be afraid of.

You got a point with the "groups comparison". We have had this discussion above (about realizing that in this example, we actually got the same enum, and there should be only one type). I think this is a tall order. Doing this at the package level would completely break the way models are currently generated.

My dev proposal (PR to come), takes it the other way around: it gives almost full control over the generated code with options and vendor some extension. This does not suits perfectly use-cases with 3rd party spec that you can't alter, but it is an improvement.

On the other hand such customization features perfectly suits the use case where you actually develop something in golang behind the API (well, my use case...).

**Comment by fredbi** [comment]
BTW, regarding enum in nested anonymous, this is currently broken (I figured that out with my latest PR).
I needed to refactor the enum validation code to see it clearer: so now it is working on dev (i.e. enum within inlined struct that you got with --skip-flatten). This fix shall come with the "enum" PR, alongside all other known bugs tagged "enum".

**Comment by gregmarr** [comment]
@fredbi Looks good to me.

>I don't know how much this is helping vs getting people to just define their own constants.

>While the code being generated here is valid, compiles and works, it probably doesn't actually capture intent. The intent being typesafe reuse, and unless we analyze all the enums and group the ones that are the same we don't capture that intent.

@casualjim My intent here is not typesafe **re**use, just plain typesafe use of a third-party spec for which I am generating a client.

**Comment by fredbi** [comment]
@gregmarr : beware this just for primitive types.
It is not much more complicated to code it for more general enum types, but it is much more complicated to test it...

**Comment by gregmarr** [comment]
@fredbi I have not yet seen `enum` used for anything other than `string` so that's fine with me right now.
